hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -104,23 +106,61 @@ public void beforeInitialize(CollectionPersister persister, int anticipatedSize)
 		this.bag = (List) persister.getCollectionType().instantiate( anticipatedSize );
 	}
 
-	@Override
-	public boolean equalsSnapshot(CollectionPersister persister) throws HibernateException {
-		final Type elementType = persister.getElementType();
-		final List sn = (List) getSnapshot();
-		if ( sn.size() != bag.size() ) {
-			return false;
-		}
-		for ( Object elt : bag ) {
-			final boolean unequal = countOccurrences( elt, bag, elementType ) != countOccurrences( elt, sn, elementType );
-			if ( unequal ) {
-				return false;
-			}
-		}
-		return true;
-	}
-
-	@Override
+    @Override
+    public boolean equalsSnapshot(CollectionPersister persister) throws HibernateException {
+        final Type elementType = persister.getElementType();
+        final List sn = (List) getSnapshot();
+        if (sn.size() != bag.size()) {
+            return false;
+        }
+
+        Map instanceCountBag = groupByEqualityHash(bag, elementType);
+        Map instanceCountSn = groupByEqualityHash(sn, elementType);
+
+        for (Object elt : bag) {
+            List candidatesBag = getItemsWithSameHash(instanceCountBag, elt, elementType);
+            List candidatesSn = getItemsWithSameHash(instanceCountSn, elt, elementType);
+
+            final boolean unequal = countOccurrences(elt, candidatesBag, elementType) 
+                    != countOccurrences(elt, candidatesSn, elementType);
+            if (unequal) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private List getItemsWithSameHash(Map instanceCountBag, Object elt, Type elementType) {
+        int hash = elementType.getHashCode(elt);
+        if (!instanceCountBag.containsKey(hash))
+            return new ArrayList();
+        else
+            return (List) instanceCountBag.get(hash);
+    }
+
+    /**
+     * Groups items in searchedBag according to persistence "equality" as
+     * defined in Type.isSame and Type.getHashCode
+     * 
+     * @param searchedBag
+     * @return Map of "eqality" hashCode to List of objects
+     */
+    private Map groupByEqualityHash(List searchedBag, Type elementType) {
+        // First group objects by equality hash
+        Map objectsByEqualityHash = new HashMap();
+
+        for (Object i : searchedBag) {
```

Comment:
I think it will be possible. I was just afraid to use JDK5+ features since this part of codebase even uses nongeneric collections (I have just noticed Hib. 5.2 requires JDK 8)